### DISCUSSION TITLE
Search for word does not run

### DISCUSSION THREAD

**Discussion by mameIIas** [pymupdf/PyMuPDF#1618] [Looking for help] [answered]
I'm trying just a simple example where a couple of words are being replaced, but I can't replace a speficic word because the program doesn't work (the words commented are the ones that broke the code) (I'm trying to find Viviana but I can't). But if I write any other word it does work, that code does run, but if I change text it does not.
What can i do?
```
import fitz

### READ IN PDF
doc = fitz.open("input.pdf")

for page in doc:
    ### SEARCH
    text = "viiana"
    #text = "En"
    #text = "fundamenta"
    text_instances = page.searchFor(text)

    ### HIGHLIGHT
    for inst in text_instances:
        highlight = page.addHighlightAnnot(inst)
        highlight.update()


### OUTPUT


doc.save("output.pdf", garbage=4, deflate=True, clean=True)
```
[test.pdf](https://github.com/pymupdf/PyMuPDF/files/8152889/test.pdf)
Update:
I've changed the words to match the file I'm uploading which has the same problem that the previous one.
And file attached.

Update 2:
fitz                               0.0.1.dev2
Python 3.9

**Comment by mameIIas** [comment]
If tried more and more, and the thing is, if the word I'm searching (text var) is in the pdf, the code does not work.

**Comment by JorjMcKie** [comment]
As I wrote: a reproducing file is required!

**Comment by JorjMcKie** [comment]
Ah, ok - I see the file now.

**Comment by JorjMcKie** [comment]
Doing this:
```python
>>> doc=fitz.open("test.pdf")
>>> page=doc[0]
>>> for r in page.search_for("en"):
	a=page.add_highlight_annot(r)
	a.update()

	
>>> doc.save("y.pdf")
>>>
```
produces this:
![grafik](https://user-images.githubusercontent.com/8290722/155976571-b44f6c96-e1a6-4349-b6a7-cd674f3b77d5.png)

Looks as it should, doesn't it?

**Comment by mameIIas** [comment]
Yes, that's what I'm looking for, but you are using "search_for" which seems deprecated for me ( 'Page' object has no attribute 'search_for') is waiting for "searchFor".
Which version are you using?

**Comment by JorjMcKie** [comment]
No, I must ask you: which version do you use?
You forgot that (mandatory!) info in your bug report.

**Comment by mameIIas** [comment]
I put it at the end of the last update ( fitz 0.0.1.dev2). The one that was installed after using pip install

**Comment by JorjMcKie** [comment]
Package "fitz" has nothing to do with PyMuPDF!

**Comment by JorjMcKie** [comment]
That package cannot coexist with PyMuPDF in the same environment.

**Comment by mameIIas** [comment]
Okay, so I delete everything of fitz and install PyMuPDF, thought that they were the same, that fitz was PyMuPDF.
Sorry and thanks for the quick answers!

**Comment by JorjMcKie** [comment]
All that could have been found earlier, if you had used a complete bug report.
You must have seen error messages opening the PDF - which you did not report!

**Comment by mameIIas** [comment]
Then I need to install Mupdf, and then pymupdf right? This is not just a python module. All this came because in everything I've found on the internet seems that fitz is the package of Pymupdf
![image](https://user-images.githubusercontent.com/34312469/155978992-51392003-55bf-4d65-a4f9-b6f9eb1c3c30.png)

**Comment by JorjMcKie** [comment, marked as answer]
You also kept your operating system secret.
If it is Mac, Linux or Windows, then there exist wheels.
If true, you only need to do this:
```python
python -m pip install -U pip
python -m pip install pymupdf
```
No separate MuPDF installation required.

**Comment by mameIIas** [comment]
I'm windows sorry

**Comment by mameIIas** [comment]
Unistalled fitz, installed pymu ( PyMuPDF 1.16.7) and upgrade pip and I keep getting the same error:
PD: I'm using anaconda
![image](https://user-images.githubusercontent.com/34312469/155981368-24a9b03d-0baa-4d89-b61f-7fc794e53134.png)

**Comment by JorjMcKie** [comment]
PyMuPDF is not in Anaconda. Use native pip as I wrote.

**Comment by mameIIas** [comment]
Yes, I used pip to install it on anaconda, that version is shown in the anaconda prompt.

**Reply by mameIIas** [reply]
Okay, the problem was my computer, tried in another one, and worked. I have an issue with the fitz package.
Thanks in advance, and sorry for bothering

**Reply by JorjMcKie** [reply]
I am always willing to help. That is no bothering.
Incrementally identifying all the required facts for a bug however is a waste of time.
So next time please use the issue templates of the main repo and try to fill them out completely.

**Reply by mameIIas** [reply]
Okay, my bad there, I know for the next time, and one last question, is it possible to replace words or have an example of it, I've seen "anonymize.py" that's kind of the way right?
Thanks in advance.

**Reply by JorjMcKie** [reply]
Let me re-iterate: You are welcome with any question!

As per replacing text:
The right approach is using so-called redaction annotations. They allow removing everything (text, links, images) that has an intersection with some given rectangle.
In addition, you can fill in new text (or just fill in some color) in the vacated rectangle. So in effect you can replace text by first removing it via a redaction and then fill in new text there.
1. Identify rectangle for text to remove
2. define a redaction annot: `page.add_redact_annot(rect, text="....", ...)`
3. Repeat steps 1 & 2 as often as needed
4. exec `page.apply_redactions(...)`

**Reply by mameIIas** [reply]
Okay, I'll give it a try and I'll keep you in touch if everything goes fine (that only depends on me(no bugs indeed)), and before starting how it works if a name starts in one line but continues in the next one?

**Reply by JorjMcKie** [reply]
> how it works if a name starts in one line but continues in the next one?

In that case, the first word part should end with a hyphen "-". (Py-) MuPDF is able to detect this, e.g. `meth-` / `od` will return rect1 containing / covering `meth` (without the hyphen) and rect2 will cover `od`.
If there is no hyphen, then the situation will be treated like normal, separate text pieces.
You can also search for text containing spaces, which may spread across more than one line. This also works - with some caveats: look up the documentation please.
